contrib_bots/lib/giphy/giphy.py

```diff
@@ -0,0 +1,115 @@
+from . import zulip_giphy_config as config
+import requests
+import logging
+import sys
+
+if config.GIPHY_API_KEY is None:
+    print('Please set up Giphy API key for this bot in zulip_giphy_config.py')
+    sys.exit(1)
+
+class GiphyHandler(object):
+    '''
+    This plugin posts a GIF in response to the keywords provided by the user.
+    Images are provided by Giphy, through the public API.
+    The bot looks for messages starting with "@giphy" or @mention of the bot
+    and responds with a message with the GIF based on provided keywords.
+    It also responds to private messages.
+    '''
+    def usage(self):
+        return '''
+            This plugin allows users to post GIFs provided by Giphy.
+            Users should preface keywords with "@giphy" or the Giphy-bot @mention.
+            The bot responds also to private messages.
+            '''
+
+    def get_url_gif_giphy(self, keyword, api_key):
+        # Return a URL for a Giphy GIF based on keywords given.
+        # In case of error, e.g. failure to fetch a GIF URL, it will
+        # return a number.
+        query = {'s': keyword,
+                 'api_key': api_key}
+        try:
+            data = requests.get('http://api.giphy.com/v1/gifs/translate', params=query)
+        except requests.exceptions.ConnectionError as e:  # Usually triggered by bad connection.
+            logging.warning(e)
+            return 1
+
+        search_status = data.json()['meta']['status']
+        if search_status != 200 or data.status_code != 200:
+            return 1
+
+        try:
+            gif_url = data.json()['data']['images']['original']['url']
+        except (TypeError, KeyError):  # Usually triggered by no result in Giphy.
+            return 2
+
+        return gif_url
+
+    def triage_message(self, message, client):
+        # To prevent infinite loop in private message, bot will detect
+        # if the sender name is the bot name it will return false.
+        if message['type'] == 'private':
+            return client.full_name != message['sender_full_name']
+
+        # Return True if we want to (possibly) response to this message.
+        original_content = message['content']
+        is_giphy_called = (original_content.startswith('@giphy ') or
+                           message['is_mentioned'])
+
+        return is_giphy_called
+
+    def handle_message(self, message, client, state_handler):
+        new_content = None
+
+        # Handle the message that called through mention.
+        if message['is_mentioned']:
+            if message['content'].startswith('@**%s** ' % (client.full_name)):
+                keyword = message['content'][len('@**%s** ' % (client.full_name)):]
+            elif message['content'].startswith('@%s ' % (client.full_name)):
+                keyword = message['content'][len('@%s ' % (client.full_name)):]
+            else:
+                new_content = 'Please mention me first, then type the keyword.'
```

Comment:
One way to structure the code here is to divide up these two jobs:
* get content (including possibly an error message)
* send content

If you extract a function called `get_new_content()`, this would allow you to do an early return for the error case and avoid the `if new_content is None` conditional.  It would also allow you to have an easily testable function that doesn't have side effects.  I would even consider putting `get_new_content()` outside of `GiphyHandler()`.

There's probably a better name than "new_content" here, too--maybe something more like "bot_response."  The variable "new_content" makes it sound like we're editing content.